File: plugins/plugin-auto-trader/rust/src/actions/compare_strategies.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::types::TradingStrategy;
use crate::{Action, ActionExample, ActionResult};

pub struct CompareStrategiesAction;

#[async_trait]
impl Action for CompareStrategiesAction {
    fn name(&self) -> &str {
        "COMPARE_STRATEGIES"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["COMPARE_BACKTEST", "STRATEGY_COMPARISON", "RANK_STRATEGIES"]
    }

    fn description(&self) -> &str {
        "Compare multiple trading strategies via backtesting."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        _message: &Value,
        state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let period = state
            .get("period_days")
            .and_then(|v| v.as_u64())
            .unwrap_or(30) as u32;

        let strategies = vec![
            TradingStrategy::Random,
            TradingStrategy::RuleBased,
            TradingStrategy::LlmDriven,
        ];

        let results = svc.compare_strategies(&strategies, period).await;

        let mut lines = vec![format!("Strategy Comparison ({} days):", period)];
        for r in &results {
            lines.push(format!(
                "  {}: PnL=${:.2}, WinRate={:.1}%, Drawdown={:.2}%, Sharpe={:.2}",
                r.strategy,
                r.total_pnl,
                r.win_rate * 100.0,
                r.max_drawdown * 100.0,
                r.sharpe_ratio,
            ));
        }

        ActionResult {
            success: true,
            text: lines.join("\n"),
            data: Some(json!(results)),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Compare all strategies over 30 days".to_string(),
            agent_response: "Strategy Comparison (30 days): ...".to_string(),
        }]
    }
}

File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/compare_strategies.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService
from elizaos_plugin_auto_trader.types import TradingStrategy


@dataclass
class CompareStrategiesAction:
    @property
    def name(self) -> str:
        return "COMPARE_STRATEGIES"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        _message: Message,
        state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        period = int(state.get("period_days", 30))
        strategies = [
            TradingStrategy.RANDOM,
            TradingStrategy.RULE_BASED,
            TradingStrategy.LLM_DRIVEN,
        ]

        results = await service.compare_strategies(strategies, period)

        lines = [f"Strategy Comparison ({period} days):"]
        for r in results:
            lines.append(
                f"  {r.strategy.value}: PnL=${r.total_pnl:.2f}, "
                f"WinRate={r.win_rate * 100:.1f}%, "
                f"Drawdown={r.max_drawdown * 100:.2f}%, "
                f"Sharpe={r.sharpe_ratio:.2f}"
            )

        return ActionResult(True, "\n".join(lines), data=results)
